File: csharp/src/api/Services/IPatientService.cs
using Example.Api.Dtos;
using Example.Api.Dtos.Requests;
using Example.Api.Dtos.Responses;

namespace Example.Api.Services;

/// <summary>
/// Service interface for managing patients.
/// </summary>
public interface IPatientService
{
    /// <summary>
    /// Retrieves patients within a specified date range.
    /// </summary>
    /// <param name="request">The request containing date range and pagination information.</param>
    /// <returns>A paged result of patients.</returns>
    Task<ApiResult<PagedResult<PatientDto>>> GetPatientsAsync(GetPatientsRequest request);

    /// <summary>
    /// Retrieves a patient by their id.
    /// </summary>
    /// <param name="id">The ID of the patient to retrieve.</param>
    /// <returns>The patient with the specified ID.</returns>
    Task<ApiResult<PatientDto>> GetPatientAsync(long id);

    /// <summary>
    /// Creates a new patient.
    /// </summary>
    /// <param name="request">The request containing patient information.</param>
    /// <returns>The created patient.</returns>
    Task<ApiResult<PatientDto>> CreatePatientAsync(CreatePatientRequest request);
}

File: src/nodejs/src/services/IPatientService.ts
import { ApiDataResult } from "../dtos/ApiResult";
import { CreatePatientRequest } from "../dtos/CreatePatientRequest";
import { PagedRequest } from "../dtos/PagedRequest";
import { PagedResult } from "../dtos/PagedResult";
import { PatientDto } from "../dtos/PatientDto";

/**
 * Interface for patient service operations.
 */
export interface IPatientService {
  /**
   * Retrieves a paginated list of patients within a date range.
   * @param request - The pagination request.
   * @param startTime - Optional start of the date range.
   * @param endTime - Optional end of the date range.
   * @returns An API result containing the paginated patient list.
   */
  getPatientsAsync(
    request: PagedRequest,
    startTime?: Date,
    endTime?: Date
  ): Promise<ApiDataResult<PagedResult<PatientDto>>>;

  /**
   * Retrieves a patient by ID.
   * @param id - The patient ID.
   * @returns An API result containing the patient if found, null otherwise.
   */
  getPatientAsync(id: number): Promise<ApiDataResult<PatientDto | null>>;

  /**
   * Creates a new patient.
   * @param request - The create patient request.
   * @returns An API result containing the created patient.
   */
  createPatientAsync(request: CreatePatientRequest): Promise<ApiDataResult<PatientDto>>;
}
